apps/src/aiTutor/types.ts

```diff
@@ -1,61 +1,68 @@
 import {
-  AiTutorInteractionSaveStatus,
-  AiTutorTypes,
+  AiTutorInteractionStatus as AITutorInteractionStatus,
+  AiTutorTypes as AITutorTypes,
+  PiiTypes as PII,
 } from '@cdo/apps/util/sharedConstants';
 
-export type ChatCompletionMessage = {
+export type AITutorTypesValue =
+  (typeof AITutorTypes)[keyof typeof AITutorTypes];
+export type AITutorInteractionStatusValue =
+  (typeof AITutorInteractionStatus)[keyof typeof AITutorInteractionStatus];
```

Comment:
I was operating at the edge of my TypeScript knowledge, so I'd appreciate a second set of eyes. I'm also tripped up by naming conventions here and would appreciate any guidance.

Reply:
I haven't seen this syntax before, but let's try to figure it out! `AITutorInteractionStatus` is
```
 AI_TUTOR_INTERACTION_STATUS = {
    ERROR: 'error',
    PII_VIOLATION: 'pii_violation',
    PROFANITY_VIOLATION: 'profanity_violation',
    OK: 'ok',
    UNKNOWN: 'unknown',
}
```
`keyof typeof AITutorInteractionStatus` gives you all the keys in an object ([helpful article](https://blog.bitsrc.io/how-to-use-keyof-typeof-in-typescript-de453fa04fef)). In this case it is `'ERROR' |  'PII_VIOLATION' |  'PROFANITY_VIOLATION' | 'OK' | 'UNKNOWN'`

`typeof AITutorInteractionStatus` is (referencing that same article) `{ERROR: string, PII_VIOLATION: string, ...}`.

If you put those together, we are getting the types for each key in `typeof AITutorInteractionStatus`, which ends up being `string | string | string ...`, which gets combined into just `string` (I tried this out locally and that's what I ended up with, assuming my initial assumption on the object was correct). If you hover over `AITutorInteractionStatusValue` do you see `string` or something else? I'm guessing we actually wanted the values of the object.

Reply:
I can't find an easy way to turn the values into a type if that's what we want.

Reply:
The `as const` assertion can be handy here - if you add that to the original object, TS treat the types as the literal values (instead of string)

Reply:
For example - 
```
const Status = {
  key1: 'value1',
  key2: 'value2',
};

// type is string
type StatusValues = (typeof Status)[keyof typeof Status];

const StatusConst = {
  key1: 'value1',
  key2: 'value2',
} as const;

// type is 'value1' | 'value2'
type StatusConstValue = (typeof StatusConst)[keyof typeof StatusConst];
```

Reply:
Thank you so much for the deep dive here, Molly and Sanchit!

Reply:
Wrinkle: The imported constant comes from here: https://github.com/code-dot-org/code-dot-org/blob/15c58936bc04b856bee226020fe636ed4e1ea667/apps/script/generateSharedConstants.rb#L40-L41. 

I tried
```
import { AiTutorTypes } from '@cdo/apps/util/sharedConstants';

const AITutorTypes = AiTutorTypes as const;
```
And that doesn't work. One idea is to modify generateSharedConstants to re-export with const assertion, but that doesn't seem worth the trouble assuming it's possible.

Reply:
Something like this doesn't seem to work either:
```
export type AITutorInteractionStatusValue =
  | AITutorInteractionStatus.ERROR
  | AITutorInteractionStatus.PII_VIOLATION
  | AITutorInteractionStatus.PROFANITY_VIOLATION
  | AITutorInteractionStatus.OK
  | AITutorInteractionStatus.UNKNOWN;
```

I'm wondering if it's just easiest to have...
```
export type AITutorInteractionStatusValue =
  | 'error'
  | 'pii_violation'
  | 'profanity_violation'
  | 'ok'
  | 'unknown';
```

Reply:
yeah I think just copying the constants over seems fine here